### ISSUE TITLE
Potential false positive for NP_NONNULL_PARAM_VIOLATION?

### ISSUE THREAD

**Issue by davidgoate** [spotbugs/spotbugs#680]
I am getting an error when running spotbugs.

# Environment

* OSX 10.13.5
* Java 10.0.1
* spotbugs-maven-plugin 3.1.5

#  Settings

```xml
<plugin>
                    <groupId>com.github.spotbugs</groupId>
                    <artifactId>spotbugs-maven-plugin</artifactId>
                    <version>3.1.5</version>
                    <configuration>
                        <effort>Max</effort>
                        <threshold>Low</threshold>
                        <excludeFilterFile>${basedir}/../spotbugs-excludes.xml</excludeFilterFile>
                        <plugins>
                            <plugin>
                                <groupId>com.h3xstream.findsecbugs</groupId>
                                <artifactId>findsecbugs-plugin</artifactId>
                                <version>1.8.0</version>
                            </plugin>
                        </plugins>
                    </configuration>
                    <executions>
                        <execution>
                            <goals>
                                <goal>check</goal>
                            </goals>
                            <phase>verify</phase>
                        </execution>
                    </executions>
                </plugin>
```

# Issue
> [ERROR] Null passed for non-null parameter of org.springframework.jdbc.core.namedparam.MapSqlParameterSource.addValue(String, Object) NP_NONNULL_PARAM_VIOLATION

I am using `spring-jdbc-5.0.7.RELEASE.jar`

Method is question is:

```java
/**
	 * Add a parameter to this parameter source.
	 * @param paramName the name of the parameter
	 * @param value the value of the parameter
	 * @return a reference to this parameter source,
	 * so it's possible to chain several calls together
	 */
	public MapSqlParameterSource addValue(String paramName, Object value) {
		Assert.notNull(paramName, "Parameter name must not be null");
		this.values.put(paramName, value);
		if (value instanceof SqlParameterValue) {
			registerSqlType(paramName, ((SqlParameterValue) value).getSqlType());
		}
		return this;
	}
```

Currently the calling code does this:

```java
if (something==null) {
 source.addValue("myValue", null);
} else {
 source.addValue("myValue", something.getMyValue());
}
```

If I invert the logic:

```java
if (something!=null) {
source.addValue("myValue", something.getMyValue());
} 
```

then i will get the exception `org.springframework.dao.InvalidDataAccessApiUsageException: No value supplied for the SQL parameter 'myValue': No value registered for key 'myValue'`

# Extra Information
My IDE (IntelliJ Ultimate  2018.1) also seems to agree that i'm passing null where i should not.

**Comment by ThrawnCA** [comment]
I don't think it's related to the conditional logic. Presumably `getMyValue()` can return `null`, which triggers the SpotBugs detector to warn you that you're potentially passing `null` to the `addValue` method, and `something` can be null, in which case you never populate "myValue" and Spring complains that it's missing.

Neither of these seems like a mistake on SpotBugs' part.

**Comment by davidgoate** [comment]
@ThrawnCA  Thanks, the part that is complaining is this line:

```java
 source.addValue("myValue", null);
```

> Null passed for non-null parameter of org.springframework.jdbc.core.namedparam.MapSqlParameterSource.addValue(String, Object).

which is inside the first part of the conditional logic (`if`). As you correctly say, the error being reported is anything to do with the conditional logic - I included that for context.  The part that it's not happy about if the hardcoded `null` being passed for the value of `addValue`.

I'd missed the `package-info.java` - which as you point out is annotated with:

```java
@NonNullApi
@NonNullFields
```
